modules/standard/Map.chpl

```diff
@@ -419,26 +419,26 @@ module Map {
         var val: valType;
         table.fillSlot(slot, k, val);
       }
-      return table.table[slot].val;
+      ref result = table.table[slot].val;
+      return result;
     }
 
-    pragma "no doc"
-    proc const this(k: keyType) const throws
-      where shouldReturnRvalueByValue(valType) &&
-            !isNonNilableClass(valType) {
+    proc ref this(k: keyType) ref throws
+      where isNonNilableClass(valType) {
       _warnForParSafeIndexing();
 
       _enter(); defer _leave();
-      var (found, slot) = table.findFullSlot(k);
-      if !found then
+
+      var (_, slot) = table.findAvailableSlot(k);
+      if !table.isSlotFull(slot) {
```

Comment:
Isn't the result that would be in the `_` spot what you're checking here?  Seems like it would make sense to check that value instead of having a separate call

Reply:
The signature here is `findAvailableSlot(key: keyType): (bool, int)` and it is returning whether or not a matching filled slot was found, not necessarily whether or not it is full, so we have to do the second `isSlotFull` check to see if that slot was full. 

So, the `slot` that is returned is the hash to the value in the hashtable, the `_` that is returned is whether or not a slot was found (it should always be found, or something went terribly wrong) and then we need to check if that slot is occupied in the table or not as a separate step.

Does that make sense?